server/src/main/java/io/spine/server/entity/storage/EntityRecordStorage.java

```diff
@@ -0,0 +1,339 @@
+import static io.spine.server.entity.storage.EntityRecordColumn.deleted;
+import static io.spine.server.entity.storage.ToEntityRecordQuery.transform;
+
+/**
+ * A storage for {@link EntityRecord}s.
+ *
+ * <p>Delegates all storage operations to the underlying
+ * {@link io.spine.server.storage.RecordStorage RecordStorage}. For that, creates a new
+ * {@code RecordStorage} instance through the provided {@code StorageFactory} and configures
+ * it with the {@linkplain EntityRecordSpec record specification} corresponding
+ * to the stored Entity.
+ *
+ * @param <I>
+ *         the type of the identifiers of stored entities
+ * @param <S>
+ *         the type of {@code Entity} state
+ */
+public class EntityRecordStorage<I, S extends EntityState<I>>
+        extends RecordStorageDelegate<I, EntityRecord> {
+
+    /**
+     * The query which aims to find only the records, which entity origins are neither archived
+     * nor deleted.
+     */
+    private final RecordQuery<I, EntityRecord> findActiveRecordsQuery;
+
+    /**
+     * Tells if the entity has {@link EntityRecordColumn#archived archived} column declared.
+     */
+    private final boolean hasArchivedColumn;
+
+    /**
+     * Tells if the entity has {@link EntityRecordColumn#deleted deleted} column declared.
+     */
+    private final boolean hasDeletedColumn;
+
+    /**
+     * Creates a new instance.
+     *
+     * <p>This constructor takes the specification of the Bounded Context in which the storage
+     * is created. It may be used by other framework libraries and SPI users to set properties
+     * of an underlying DBMS (such as DB table names), with which this record storage interacts.
+     *
+     * @param context specification of the Bounded Context in which the created storage will be used
+     * @param factory storage factory
+     * @param entityClass class of an Entity which data is stored in the created storage
+     */
+    public EntityRecordStorage(ContextSpec context,
+                               StorageFactory factory,
+                               Class<? extends Entity<I, S>> entityClass) {
+        super(context, factory.createRecordStorage(context, spec(entityClass)));
+        this.hasArchivedColumn = hasColumn(archived);
+        this.hasDeletedColumn = hasColumn(deleted);
+        FindActiveEntites<I, S> entityQuery = findActiveEntities().build();
+        this.findActiveRecordsQuery = transform(entityQuery);
+    }
+
+    private static <I, S extends EntityState<I>> EntityRecordSpec<I, S, ?>
+    spec(Class<? extends Entity<I, S>> entityClass) {
+        return EntityRecordSpec.of(entityClass);
+    }
+
+    /**
+     * Returns the iterator over all non-archived and non-deleted entity records.
+     *
+     * @throws IllegalStateException
+     *         if the storage is already closed
+     */
+    @Override
+    public Iterator<I> index() {
+        Iterator<EntityRecord> iterator = super.readAll(findActiveRecordsQuery);
```

Comment:
This is definitely not for this PR, but we need to make it more efficient. It looks like we're reading everything just to get all the identifiers.